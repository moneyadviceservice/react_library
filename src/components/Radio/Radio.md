The Radio input component allows the user to enable or disable options.

```jsx
import React, { useState } from 'react';
import { Form, Formfield, Radiobutton, Paragraph as P } from '@moneypensionservice/directories';

const [fruit, changeValue] = useState('Mango');

<>
<Form>
  <Formfield legend="Choose your favourite fruit:" label="Fruits">
    <Radio
      checked={fruit === 'Pineapple'}
      onChange={e => changeValue(e.target.value)}
      value="Pineapple"
      label="🍍 Pineapple"
      name="fruits" />
    <Radio
      checked={fruit === 'Mango'}
      onChange={e => changeValue(e.target.value)}
      value="Mango"
      label="🥭 Mango"
      name="fruits" />
    <Radio
      checked={fruit === 'Melon'}
      onChange={e => changeValue(e.target.value)}
      value="Melon"
      label="🍈 Melon"
      name="fruits" />
    <Radio
      checked={fruit === 'Banana'}
      onChange={e => changeValue(e.target.value)}
      value="Banana"
      label="🍌 Banana"
      name="fruits"
      disabled />
  </Formfield>
</Form>
<P margin={{top: '10px'}}>{`Currently selected: ${fruit}`}</P>
</>
```
