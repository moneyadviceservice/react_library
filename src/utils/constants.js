export const breakpoints = {
  xs: 1,
  sm: 48, // 768px
  md: 64, // 1024px
  lg: 90, // 1440px
  xl: 120, // 1920px
}

export const dimensions = ['xs', 'sm', 'md', 'lg', 'xl']

export const mediaQuery = 'only screen'

export const ALIGN_SELF_MAP = {
  center: 'center',
  end: 'flex-end',
  start: 'flex-start',
  stretch: 'stretch',
}

export const SPACING_SIZES = [
  'xxsmall',
  'xsmall',
  'small',
  'medium',
  'large',
  'xlarge',
]

