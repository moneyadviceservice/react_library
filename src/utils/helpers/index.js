export * from './background'
export * from './responsive'
export * from './flexbox'
export * from './styles'
export * from './mediaQueries'
export * from './sizing'
export * from './refs'
