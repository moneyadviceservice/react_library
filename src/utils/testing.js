import { render, act } from '@testing-library/react'
import GlobalTheme from '../components/ThemeProvider'

const customRender = (ui, options) =>
  render(ui, { wrapper: GlobalTheme, ...options })

const waitRender = async (ui, options) => {
  const { container, ...rest } = customRender(ui, options)

  await act(async () => {
    await Promise.resolve(container)
  })

  return { container, ...rest }
}

export * from '@testing-library/react'

export { customRender as render, waitRender }
