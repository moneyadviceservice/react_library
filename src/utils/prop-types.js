import PropTypes from 'prop-types'
import { SPACING_SIZES } from './constants'

/** spacing prop.type */
const spacingProp = PropTypes.shape({
  bottom: PropTypes.oneOfType([
    PropTypes.oneOf(SPACING_SIZES),
    PropTypes.string,
  ]),
  horizontal: PropTypes.oneOfType([
    PropTypes.oneOf(SPACING_SIZES),
    PropTypes.string,
  ]),
  left: PropTypes.oneOfType([PropTypes.oneOf(SPACING_SIZES), PropTypes.string]),
  right: PropTypes.oneOfType([
    PropTypes.oneOf(SPACING_SIZES),
    PropTypes.string,
  ]),
  top: PropTypes.oneOfType([PropTypes.oneOf(SPACING_SIZES), PropTypes.string]),
  vertical: PropTypes.oneOfType([
    PropTypes.oneOf(SPACING_SIZES),
    PropTypes.string,
  ]),
})

export const genericPropTypes = {
  a11yTitle: PropTypes.string,
  alignSelf: PropTypes.oneOf(['start', 'center', 'end', 'stretch']),
  margin: PropTypes.oneOfType([
    PropTypes.oneOf(['none', ...SPACING_SIZES]),
    spacingProp,
    PropTypes.string,
    PropTypes.object,
  ]),
  padding: PropTypes.oneOfType([
    PropTypes.oneOf([...SPACING_SIZES]),
    spacingProp,
    PropTypes.string,
    PropTypes.object,
  ]),
  border: PropTypes.oneOfType([
    PropTypes.oneOf(['none', ...SPACING_SIZES]),
    spacingProp,
    PropTypes.string,
    PropTypes.object,
  ]),
}

export function genericPropsDefaults({
  a11yTitle = null,
  alignSelf = null,
  margin = null,
  padding = null,
  border = null,
} = {}) {
  return {
    a11yTitle: a11yTitle,
    alignSelf: alignSelf,
    margin: margin,
    padding: padding,
    border: border,
  }
}
