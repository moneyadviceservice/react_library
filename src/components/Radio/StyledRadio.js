import styled from 'styled-components'
import { genericStyles } from '../../utils/helpers'

const Label = styled.label`
  ${genericStyles}
  ${({ disabled, theme }) => disabled && `color: ${theme.colors.disabledText};`}
  ${({ hide }) => !hide && `display: block;`}
`

const Field = styled.input`
  margin-right: 5px;
  padding: 0;
`

export { Label, Field }
