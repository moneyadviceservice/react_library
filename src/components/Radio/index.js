import React from 'react'
import PropTypes from 'prop-types'
import { genericPropTypes, genericPropsDefaults } from '../../utils/prop-types'
import { Label, Field } from './StyledRadio'
import { Paragraph as P } from '../Paragraph'

const Radio = ({
  a11yTitle,
  checked,
  children,
  disabled,
  label,
  name,
  onChange,
  value,
  ...rest
}) => (
  <Label disabled={disabled} {...rest}>
    <Field
      aria-label={a11yTitle}
      checked={checked}
      disabled={disabled}
      name={name}
      onChange={onChange}
      type="radio"
      value={value}
    />
    <P textSize="1rem" margin="0">
      {label}
    </P>
    {children}
  </Label>
)

// Documentation
Radio.propTypes = {
  /** Whether the component is checked. */
  checked: PropTypes.bool,
  /** Text inside the Label component. Works as the caption for the checkbox element. */
  children: PropTypes.string,
  /** Whether the form control is disabled. */
  disabled: PropTypes.bool,
  /** Caption for the checkbox element. */
  label: PropTypes.string,
  /** Name of the input form control. Submitted with the form as part of a name/value pair. */
  name: PropTypes.string,
  /** On change trigger function event. */
  onChange: PropTypes.func,
  /** Current value of the form control. Submitted with the form as part of a name/value pair. */
  value: PropTypes.string,
  ...genericPropTypes,
}

Radio.defaultProps = {
  disabled: false,
  ...genericPropsDefaults(),
}

/** @component */
export { Radio }
