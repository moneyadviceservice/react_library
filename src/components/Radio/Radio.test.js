import React from 'react'
import { render, fireEvent } from '../../utils/testing'
import { Radio } from '../Radio'

describe('<Radio />', () => {
  describe('Snapshots', () => {
    it('should match snapshot without props', () => {
      const { container } = render(<Radio />)
      expect(container).toMatchSnapshot()
    })
    it('should match snapshot with disabled prop', () => {
      const { container } = render(<Radio disabled />)
      expect(container).toMatchSnapshot()
    })

    it('should match snapshot with checked prop', () => {
      const { container } = render(<Radio checked onChange={e => {}} />)
      expect(container).toMatchSnapshot()
    })

    it('should match snapshot with checked and disabled prop', () => {
      const { container } = render(
        <Radio checked disabled onChange={e => {}} />
      )
      expect(container).toMatchSnapshot()
    })
  })

  describe('onChange prop', () => {
    it('should call onChange callback when RadioButton is changed', () => {
      const onChangeMock = jest.fn()
      const { getByTestId } = render(
        <Radio onChange={onChangeMock} data-testid="radio">
          Radio
        </Radio>
      )

      fireEvent.click(getByTestId('radio'))

      expect(onChangeMock).toHaveBeenCalledTimes(1)
    })
  })
})
