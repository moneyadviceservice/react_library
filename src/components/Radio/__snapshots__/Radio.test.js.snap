// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Radio /> Snapshots should match snapshot with checked and disabled prop 1`] = `
<div>
  <label
    class="StyledRadio__Label-sc-1awvsvp-0 fDTlvj"
    disabled=""
  >
    <input
      checked=""
      class="StyledRadio__Field-sc-1awvsvp-1 bBSLqf"
      disabled=""
      type="radio"
      value=""
    />
    <p
      class="StyledParagraph-sc-1toaej-0 ZKQVH"
    />
  </label>
</div>
`;

exports[`<Radio /> Snapshots should match snapshot with checked prop 1`] = `
<div>
  <label
    class="StyledRadio__Label-sc-1awvsvp-0 fYWYHJ"
  >
    <input
      checked=""
      class="StyledRadio__Field-sc-1awvsvp-1 bBSLqf"
      type="radio"
      value=""
    />
    <p
      class="StyledParagraph-sc-1toaej-0 ZKQVH"
    />
  </label>
</div>
`;

exports[`<Radio /> Snapshots should match snapshot with disabled prop 1`] = `
<div>
  <label
    class="StyledRadio__Label-sc-1awvsvp-0 fDTlvj"
    disabled=""
  >
    <input
      class="StyledRadio__Field-sc-1awvsvp-1 bBSLqf"
      disabled=""
      type="radio"
      value=""
    />
    <p
      class="StyledParagraph-sc-1toaej-0 ZKQVH"
    />
  </label>
</div>
`;

exports[`<Radio /> Snapshots should match snapshot without props 1`] = `
<div>
  <label
    class="StyledRadio__Label-sc-1awvsvp-0 fYWYHJ"
  >
    <input
      class="StyledRadio__Field-sc-1awvsvp-1 bBSLqf"
      type="radio"
      value=""
    />
    <p
      class="StyledParagraph-sc-1toaej-0 ZKQVH"
    />
  </label>
</div>
`;
